File: ipcalc.go
// ipcalc.go
package main

import (
	"encoding/json"
	"flag"
	"fmt"
	"strconv"
	"strings"
)

func ipToInt(ip string) uint32 {
	parts := strings.Split(ip, ".")
	return (uint32(parseInt(parts[0])) << 24) |
		(uint32(parseInt(parts[1])) << 16) |
		(uint32(parseInt(parts[2])) << 8) |
		uint32(parseInt(parts[3]))
}

func intToIP(n uint32) string {
	return fmt.Sprintf("%d.%d.%d.%d", (n>>24)&0xff, (n>>16)&0xff, (n>>8)&0xff, n&0xff)
}

func parseInt(s string) int {
	v, _ := strconv.Atoi(s)
	return v
}

func prefixToMask(prefix int) uint32 {
	return ^uint32(0) << (32 - prefix)
}

func maskToPrefix(mask uint32) int {
	return bitsOn(mask)
}

func bitsOn(n uint32) int {
	count := 0
	for n != 0 {
		count++
		n &= n - 1
	}
	return count
}

func ipBinary(n uint32) string {
	return fmt.Sprintf("%08b.%08b.%08b.%08b", (n>>24)&0xff, (n>>16)&0xff, (n>>8)&0xff, n&0xff)
}

type IPCalc struct {
	ipInt      uint32
	maskInt    uint32
	prefix     int
	networkInt uint32
	broadcast  uint32
	wildcard   uint32
	first      uint32
	last       uint32
	total      uint32
	usable     uint32
}

func NewIPCalc(input, maskStr string) (*IPCalc, error) {
	var ipInt, maskInt uint32
	var prefix int
	if strings.Contains(input, "/") {
		parts := strings.Split(input, "/")
		ipInt = ipToInt(parts[0])
		prefix, _ = strconv.Atoi(parts[1])
		maskInt = prefixToMask(prefix)
	} else if maskStr != "" {
		ipInt = ipToInt(input)
		if strings.Contains(maskStr, ".") {
			maskInt = ipToInt(maskStr)
			prefix = maskToPrefix(maskInt)
		} else {
			prefix, _ = strconv.Atoi(maskStr)
			maskInt = prefixToMask(prefix)
		}
	} else {
		return nil, fmt.Errorf("invalid input")
	}
	c := &IPCalc{
		ipInt:   ipInt,
		maskInt: maskInt,
		prefix:  prefix,
	}
	c.networkInt = ipInt & maskInt
	c.broadcast = c.networkInt | (^maskInt & 0xffffffff)
	c.wildcard = ^maskInt & 0xffffffff
	if c.networkInt == c.broadcast {
		c.first = 0
		c.last = 0
	} else {
		c.first = c.networkInt + 1
		c.last = c.broadcast - 1
	}
	c.total = 1 << (32 - prefix)
	c.usable = c.total
	if c.total > 2 {
		c.usable = c.total - 2
	}
	return c, nil
}

func (c *IPCalc) toMap() map[string]interface{} {
	m := map[string]interface{}{
		"ip":           intToIP(c.ipInt),
		"mask":         intToIP(c.maskInt),
		"prefix":       c.prefix,
		"network":      intToIP(c.networkInt),
		"broadcast":    intToIP(c.broadcast),
		"wildcard":     intToIP(c.wildcard),
		"total_hosts":  c.total,
		"usable_hosts": c.usable,
	}
	if c.first != 0 {
		m["first_usable"] = intToIP(c.first)
		m["last_usable"] = intToIP(c.last)
	} else {
		m["first_usable"] = nil
		m["last_usable"] = nil
	}
	return m
}

func (c *IPCalc) subnets(newPrefix int) ([]map[string]string, error) {
	if newPrefix <= c.prefix {
		return nil, fmt.Errorf("new prefix must be larger than original")
	}
	num := 1 << (newPrefix - c.prefix)
	step := uint32(1 << (32 - newPrefix))
	var subnets []map[string]string
	for i := 0; i < num; i++ {
		net := c.networkInt + uint32(i)*step
		subCalc, _ := NewIPCalc(intToIP(net), strconv.Itoa(newPrefix))
		first := "none"
		last := "none"
		if subCalc.first != 0 {
			first = intToIP(subCalc.first)
			last = intToIP(subCalc.last)
		}
		subnets = append(subnets, map[string]string{
			"network":      intToIP(subCalc.networkInt),
			"prefix":       strconv.Itoa(subCalc.prefix),
			"first_usable": first,
			"last_usable":  last,
		})
	}
	return subnets, nil
}

func main() {
	var (
		input   = flag.String("input", "", "IP address (CIDR or with mask)")
		mask    = flag.String("mask", "", "Subnet mask if not in CIDR")
		subnet  = flag.Int("subnet", 0, "Create subnets of this prefix length")
		binary  = flag.Bool("bin", false, "Show binary")
		jsonOut = flag.Bool("json", false, "Output JSON")
	)
	flag.Parse()
	if *input == "" && len(flag.Args()) > 0 {
		*input = flag.Args()[0]
	}
	if len(flag.Args()) > 1 && *mask == "" {
		*mask = flag.Args()[1]
	}
	if *input == "" {
		fmt.Println("Usage: ipcalc <ip>[/prefix] [mask] [options]")
		return
	}
	calc, err := NewIPCalc(*input, *mask)
	if err != nil {
		fmt.Printf("Error: %v\n", err)
		return
	}
	res := calc.toMap()
	if *binary {
		res["binary_ip"] = ipBinary(calc.ipInt)
		res["binary_mask"] = ipBinary(calc.maskInt)
		res["binary_network"] = ipBinary(calc.networkInt)
		res["binary_broadcast"] = ipBinary(calc.broadcast)
	}
	var subnets []map[string]string
	if *subnet > 0 {
		subnets, err = calc.subnets(*subnet)
		if err != nil {
			fmt.Printf("Subnet error: %v\n", err)
			return
		}
	}
	if *jsonOut {
		out := res
		if len(subnets) > 0 {
			out["subnets"] = subnets
		}
		b, _ := json.MarshalIndent(out, "", "  ")
		fmt.Println(string(b))
		return
	}
	fmt.Println("\nIP Calculator")
	fmt.Printf("IP Address:     %s\n", res["ip"])
	fmt.Printf("Subnet Mask:    %s (/%d)\n", res["mask"], res["prefix"])
	fmt.Printf("Network:        %s\n", res["network"])
	fmt.Printf("Broadcast:      %s\n", res["broadcast"])
	fmt.Printf("Wildcard Mask:  %s\n", res["wildcard"])
	if res["first_usable"] != nil {
		fmt.Printf("First Usable:   %s\n", res["first_usable"])
		fmt.Printf("Last Usable:    %s\n", res["last_usable"])
	} else {
		fmt.Println("First Usable:   (none)")
		fmt.Println("Last Usable:    (none)")
	}
	fmt.Printf("Total Hosts:    %d\n", res["total_hosts"])
	fmt.Printf("Usable Hosts:   %d\n", res["usable_hosts"])
	if *binary {
		fmt.Println("\nBinary:")
		fmt.Printf("IP:      %s\n", res["binary_ip"])
		fmt.Printf("Mask:    %s\n", res["binary_mask"])
		fmt.Printf("Network: %s\n", res["binary_network"])
	}
	if len(subnets) > 0 {
		fmt.Printf("\nSubnets (/%d):\n", *subnet)
		for _, s := range subnets {
			fmt.Printf("  %s/%s   (%s - %s)\n", s["network"], s["prefix"], s["first_usable"], s["last_usable"])
		}
	}
}

File: ipcalc.py
# ipcalc.py
import sys
import json
import argparse
import re
from typing import Tuple, List, Optional

def ip_to_int(ip: str) -> int:
    parts = ip.split('.')
    return (int(parts[0]) << 24) | (int(parts[1]) << 16) | (int(parts[2]) << 8) | int(parts[3])

def int_to_ip(num: int) -> str:
    return f"{(num >> 24) & 0xff}.{(num >> 16) & 0xff}.{(num >> 8) & 0xff}.{num & 0xff}"

def prefix_to_mask(prefix: int) -> int:
    return (~0 << (32 - prefix)) & 0xffffffff

def mask_to_prefix(mask: int) -> int:
    return bin(mask).count('1')

def ip_binary(ip: int) -> str:
    return '.'.join(f"{((ip >> (24 - i*8)) & 0xff):08b}" for i in range(4))

class IPCalc:
    def __init__(self, ip_str: str, mask_str: Optional[str] = None):
        if '/' in ip_str:
            ip_part, prefix_str = ip_str.split('/')
            self.ip_int = ip_to_int(ip_part)
            self.prefix = int(prefix_str)
            self.mask_int = prefix_to_mask(self.prefix)
        elif mask_str:
            self.ip_int = ip_to_int(ip_str)
            if '.' in mask_str:
                self.mask_int = ip_to_int(mask_str)
                self.prefix = mask_to_prefix(self.mask_int)
            else:
                self.prefix = int(mask_str)
                self.mask_int = prefix_to_mask(self.prefix)
        else:
            raise ValueError("Invalid input")
        self.ip_str = int_to_ip(self.ip_int)
        self.mask_str = int_to_ip(self.mask_int)
        self.network_int = self.ip_int & self.mask_int
        self.broadcast_int = self.network_int | (~self.mask_int & 0xffffffff)
        self.wildcard_int = ~self.mask_int & 0xffffffff
        self.first_usable = self.network_int + 1 if self.network_int != self.broadcast_int else None
        self.last_usable = self.broadcast_int - 1 if self.network_int != self.broadcast_int else None
        self.total_hosts = 1 << (32 - self.prefix)
        self.usable_hosts = max(0, self.total_hosts - 2)

    def to_dict(self) -> dict:
        return {
            "ip": self.ip_str,
            "mask": self.mask_str,
            "prefix": self.prefix,
            "network": int_to_ip(self.network_int),
            "broadcast": int_to_ip(self.broadcast_int),
            "wildcard": int_to_ip(self.wildcard_int),
            "first_usable": int_to_ip(self.first_usable) if self.first_usable else None,
            "last_usable": int_to_ip(self.last_usable) if self.last_usable else None,
            "total_hosts": self.total_hosts,
            "usable_hosts": self.usable_hosts,
            "binary_ip": ip_binary(self.ip_int),
            "binary_mask": ip_binary(self.mask_int)
        }

    def subnet(self, new_prefix: int) -> List[dict]:
        if new_prefix <= self.prefix:
            raise ValueError("New prefix must be larger than original")
        num_subnets = 1 << (new_prefix - self.prefix)
        step = 1 << (32 - new_prefix)
        subnets = []
        for i in range(num_subnets):
            net = self.network_int + i * step
            sub = IPCalc(int_to_ip(net), str(new_prefix))
            subnets.append({
                "network": sub.ip_str,
                "mask": sub.mask_str,
                "prefix": sub.prefix,
                "broadcast": sub.broadcast_str,
                "first_usable": sub.first_usable_str,
                "last_usable": sub.last_usable_str
            })
        return subnets

    @property
    def broadcast_str(self): return int_to_ip(self.broadcast_int)
    @property
    def first_usable_str(self): return int_to_ip(self.first_usable) if self.first_usable else None
    @property
    def last_usable_str(self): return int_to_ip(self.last_usable) if self.last_usable else None

def main():
    parser = argparse.ArgumentParser(description="IP Calculator")
    parser.add_argument('input', help="IP address (e.g., 192.168.1.0/24 or 192.168.1.10 255.255.255.0)")
    parser.add_argument('mask', nargs='?', help="Subnet mask if not in CIDR")
    parser.add_argument('--subnet', type=int, help="Create subnets of this prefix length")
    parser.add_argument('--bin', action='store_true', help="Show binary representation")
    parser.add_argument('--json', action='store_true', help="Output as JSON")
    args = parser.parse_args()

    try:
        calc = IPCalc(args.input, args.mask)
    except Exception as e:
        print(f"Error: {e}", file=sys.stderr)
        sys.exit(1)

    result = calc.to_dict()
    if args.bin:
        result['binary_network'] = ip_binary(calc.network_int)
        result['binary_broadcast'] = ip_binary(calc.broadcast_int)

    subnets = []
    if args.subnet:
        try:
            subnets = calc.subnet(args.subnet)
        except ValueError as e:
            print(f"Subnet error: {e}", file=sys.stderr)
            sys.exit(1)

    if args.json:
        output = result.copy()
        if subnets:
            output['subnets'] = subnets
        print(json.dumps(output, indent=2))
    else:
        print("\nIP Calculator")
        print(f"IP Address:     {result['ip']}")
        print(f"Subnet Mask:    {result['mask']} (/{result['prefix']})")
        print(f"Network:        {result['network']}")
        print(f"Broadcast:      {result['broadcast']}")
        print(f"Wildcard Mask:  {result['wildcard']}")
        if result['first_usable']:
            print(f"First Usable:   {result['first_usable']}")
            print(f"Last Usable:    {result['last_usable']}")
        else:
            print("First Usable:   (none)")
            print("Last Usable:    (none)")
        print(f"Total Hosts:    {result['total_hosts']}")
        print(f"Usable Hosts:   {result['usable_hosts']}")

        if args.bin:
            print("\nBinary:")
            print(f"IP:      {result['binary_ip']}")
            print(f"Mask:    {result['binary_mask']}")
            print(f"Network: {ip_binary(calc.network_int)}")

        if subnets:
            print(f"\nSubnets (/{args.subnet}):")
            for sub in subnets:
                first = sub['first_usable'] or "none"
                last = sub['last_usable'] or "none"
                print(f"  {sub['network']}/{sub['prefix']}   ({first} - {last})")

if __name__ == "__main__":
    main()
